build/source/engine/allocspace.f90

```diff
@@ -107,35 +109,55 @@ subroutine allocate_gru_struc(nGRU,nHRU,err,message)
+   else
+    err=1; message=trim(message)//'startGRU is larger than then GRU dimension'; return
+   end if
+  end if 
+ end if 
  
- allocate(gru_struc(nGRU), index_map(nHRU), stat=err)
- if(err/=0)then; err=20; message=trim(message)//'problem allocating space for mapping structures'; return; endif
+ ! get hru_dim dimension length (ie., the number of hrus in entire domain)
+ err = nf90_inq_dimid(ncid, "hru", hruDimID);                if(err/=nf90_noerr)then; message=trim(message)//'problem finding hru dimension/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_inquire_dimension(ncid, hruDimID, len = maxHRU); if(err/=nf90_noerr)then; message=trim(message)//'problem reading hru dimension/'//trim(nf90_strerror(err)); return; endif  
  
- allocate(gru_id(nGRU),hru_id(nHRU),hru2gru_id(nHRU),stat=err)
- if(err/=0)then; err=20; message=trim(message)//'problem allocating space for zLocalAttributes gru-hru correspondence vectors'; return; endif
-
- ! read gru_id from netcdf file
- err = nf90_inq_varid(ncid, "gruId", varid);     if(err/=0)then; message=trim(message)//'problem finding gruId variable'; return; endif
- err = nf90_get_var(ncid,varid,gru_id);          if(err/=0)then; message=trim(message)//'problem reading gruId variable'; return; endif
-
  ! read the HRU to GRU mapping
- err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=0)then; message=trim(message)//'problem finding hru2gruId variable'; return; endif
- err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=0)then; message=trim(message)//'problem reading hru2gruId variable'; return; endif
-
- ! allocate mapping array
- do iGRU=1,nGRU
-  hruCount = count(hru2gru_id==gru_id(iGRU))
-  gru_struc(iGRU)%hruCount = hruCount
-  allocate(gru_struc(iGRU)%hruInfo(hruCount),stat=err)
-  if(err/=0)then; err=20; message=trim(message)//'problem allocating space for hru in gru_struc'; return; endif
- enddo
-
+ allocate(hru2gru_id(maxHRU),stat=err)
+ if(err/=0)then; err=20; message=trim(message)//'problem allocating space for zLocalAttributes gru-hru correspondence vectors/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=nf90_noerr)then; message=trim(message)//'problem finding hru2gruId variable/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=nf90_noerr)then; message=trim(message)//'problem reading hru2gruId variable/'//trim(nf90_strerror(err)); return; endif
+ 
+ ! allocate mapping array 
+ allocate(gru_struc(nGRU))  
+ if (nHRU == integerMissing) then
+  ! get GRU Id
+  allocate(gru_id(nGRU))  
+  err = nf90_inq_varid(ncid, "gruId", varid);                              if(err/=nf90_noerr)then; message=trim(message)//'problem finding gruId variable/'//trim(nf90_strerror(err)); return; endif
+  err = nf90_get_var(ncid,varid,gru_id,start=(/startGRU/),count=(/nGRU/)); if(err/=nf90_noerr)then; message=trim(message)//'problem reading gruId variable/'//trim(nf90_strerror(err)); return; endif  
```

Comment:
This assumes we're just going to grab nGRU from the file in order?   What if I wanted to run GRU 1, 20, and 25, and they are in the netcdf files at locations 1, 40, and 3?  Are we going to require the user to have the GRUs in the order they want to run them?  Or they'll just have to figure out where they are in the netcdf files and launch SUMMA jobs accordingly?